src/SqlStreamStore.Postgres/PostgresStreamStore.ReadAll.cs

```diff
@@ -2,162 +2,371 @@
     using System.Data.Common;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Text;
     using System.Threading;
     using System.Threading.Tasks;
+    using Npgsql;
+    using SqlStreamStore.Logging;
     using SqlStreamStore.PgSqlScripts;
     using SqlStreamStore.Streams;
 
     partial class PostgresStreamStore
     {
-        protected override async Task<ReadAllPage> ReadAllForwardsInternal(
-            long fromPositionExclusive,
+        protected override async Task<ReadAllPage> ReadAllForwardsInternal(long fromPositionExclusive, int maxCount, bool prefetch, ReadNextAllPage readNext, CancellationToken cancellationToken)
+        {
+            var correlation = Guid.NewGuid();
+
+            maxCount = maxCount == int.MaxValue ? maxCount - 1 : maxCount;
+
+            var (messages, maxAgeDict, transactionIdsInProgress, isEnd) = await ReadAllForwards(fromPositionExclusive, maxCount, prefetch, correlation, cancellationToken).ConfigureAwait(false);
+
+            if(_settings.GapHandlingSettings != null)
+            {
+                var r = await HandleGaps(messages, maxAgeDict, transactionIdsInProgress, isEnd, fromPositionExclusive, maxCount, prefetch, correlation, cancellationToken).ConfigureAwait(false);
+
+                isEnd = r.isEnd;
+                messages = r.messages;
+            }
+
+            if(!messages.Any())
+            {
+                return new ReadAllPage(fromPositionExclusive, fromPositionExclusive, isEnd, ReadDirection.Forward, readNext, Array.Empty<StreamMessage>());
+            }
+
+            var filteredMessages = FilterExpired(messages, maxAgeDict);
+            var nextPosition = filteredMessages[filteredMessages.Count - 1].Position + 1;
+
+            return new ReadAllPage(fromPositionExclusive, nextPosition, isEnd, ReadDirection.Forward, readNext, filteredMessages.ToArray());
+        }
+
+        private async Task<(ReadOnlyCollection<StreamMessage> messages, ReadOnlyDictionary<string, int> maxAgeDict, bool isEnd)> HandleGaps(
+            ReadOnlyCollection<StreamMessage> messages,
+            ReadOnlyDictionary<string, int> maxAgeDict,
+            TxIdList transactionsInProgress,
+            bool isEnd,
+            long fromPositionInclusive,
             int maxCount,
             bool prefetch,
-            ReadNextAllPage readNext,
+            Guid correlation,
             CancellationToken cancellationToken)
         {
-            maxCount = maxCount == int.MaxValue ? maxCount - 1 : maxCount;
+            var hasMessages = messages.Count > 0;
+            var hasTransactionsInProgress = transactionsInProgress.Count > 0;
+            
+            Logger.TraceFormat("Correlation: {correlation} | {messages} | Transactions in progress: {transactions}",
+                correlation,
+                hasMessages ? $"Count: {messages.Count} | {string.Join("|", messages.Select((x, i) => $"Position: {x.Position} Array index: {i}"))}" : "No messages",
```

Comment:
Final opinion now you know it adds 10s :D

![image](https://user-images.githubusercontent.com/6287467/201854318-b91cbc31-1edf-4ebb-894b-b3395b132e49.png)